# ISTM 6212 - Goup Project 04 
Group members: 
* Jiaru Xu
* Min Yang 
* Jiaying Liu
* Congying Ma


# Part 1 - Selection (30 points)

Identify and describe your dataset, its source, and what appeals to you about it. Acquire the data and perform an initial exploration to determine which themes you wish to explore. Describe the questions you want to be able to answer with the data, any concerns you have about the data, and any challenges you expect to have to overcome.

## What dataset do we choose?

We get our dataset from the website https://controllerdata.lacity.org/Payroll/City-Employee-Payroll/pazn-qyym through the website provided by professor http://us-cities.survey.okfn.org.  Our dataset records all Los Angeles City employees’ payrolls from 2013 to 2018. The original dataset contains 370k records and 35 columns. It has a good mix of quantitative and qualitative variable for us to analyze.  We are specifically interested in this dataset because it includes quarterly payment and has a good deviation of columns such as base pay, permanent bonus pay, and overtime pay that we can build an analysis on how these extra payments affect the total payment, and if we can draw a job type characteristics from this analysis. The year column would also allow us to make a timeline analysis, so this dataset provides a good amount of information and large enough data for us to analyze from several perspectives. 

## What topic we wish to explore?

Since this is a dataset about city employee payroll, our major variable will be the payment. We want to explore the character of payment.  We come up with two focuses, first, we want to explore the relationship between payment and year, payment and department, and payment and job title. From there, we can break down to a specific department and if there is a unique pattern of payment for a certain department. The other focus is we are interested in finding if there is a seasonality pattern in the quarterly payment. Since most of the departments are service-based, such as Fire department and Police department, we are specifically interested in seeing if there is a difference in their payment from different quarters. 



## What questions are we going to answer?

For our first focus, we want to answer the following questions: 
1.    There should be a slowly increasing in salary over years due to inflation and living expense increase, but there might also not change a lot due to increasing demand. Does year really have an effect on salary?
2.    What are the departments with the highest average total payment and if there is something in common through these departments? The most common city employees we know for daily are fireman and policeman, since they are service-based job, we think they probably have higher payments.
3.    We know that total payment is generated from base pay, overtime pay, permanent bonus pay, etc., we want to see if these bonus pays will make a difference in total pay?
4.    We predict that job title has an impact on salary, as people with higher positions will have a higher pay, but is there a big difference between different departments? We would want to make a comparison here.
5.    Our dataset also contains employment type, we should get a conclusion that full-time gets to pay most, but we probably would find some interesting thing if we categorize each employment type and the group of departments appears to have the highest pay might be different.

For our second focus, we want to answer the following questions:
1.    We want to see if there is a difference in quarterly payment. We predict that for different department, there is a different pattern due to the department characteristics. 
2.    Second, for different positions in some specific department, we wonder if there is also a pattern over how employees are rewarded through the years in different quarters. We predict that in different quarters of the year, there exists some variation in the pattern due to the seasonality and the feature of a specific department.
3.    Third, if we find out some seasonal difference after the previous step, we would like to explore further into the reason behind and find out how the variation differs among various positions in the same department and how large the variation is. 


## Some concerns we have about the data

The dataset only contains payment, so we are lack on the employee’s personal information, such as age, gender, and years of service. With these extra columns, when we analyze some questions listed above, we probably could get a more precise explanation/reason of why certain pattern occurs. And it would also be interesting to see if same job title’s payment would be different with individual differences. 
Since q4 of 2018 is not recorded yet, it might have some impact when we analyze on years and quarterly, so we would want to treat it carefully when we come to these analyses, probably just exclude this year. 

## Challenges we expect to have to overcome.

One concern would be if we could not find there are differences, we would need to find a new dataset. 
Since our dataset has a lot of columns, and none of us are familiar to payroll, we probably will have a hard time when wrangling the dataset, to narrow down the columns we want to analyze, and if there appears to have a lot missing data in one column which might affect our analysis. For a better understanding of certain departments that appear in our rank and we don’t know, we get our department information from https://www.lacity.org/your-government/departments-commissions/departments-bureaus. 

Extra: Further analysis would require more information, it would be helpful if we can get the quarterly overtime pay which would probably strengthen the conclusion we made here. 


---

### Take a first look at our original dataset

In [1]:
!wget -O City_Employee_Payroll.csv https://controllerdata.lacity.org/api/views/pazn-qyym/rows.csv

--2018-12-10 16:27:38--  https://controllerdata.lacity.org/api/views/pazn-qyym/rows.csv
Resolving controllerdata.lacity.org (controllerdata.lacity.org)... 52.206.140.205, 52.206.140.199, 52.206.68.26
Connecting to controllerdata.lacity.org (controllerdata.lacity.org)|52.206.140.205|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/csv]
Saving to: ‘City_Employee_Payroll.csv’

City_Employee_Payro     [           <=>      ] 110.30M  50.9MB/s    in 2.2s    

Last-modified header invalid -- time-stamp ignored.
2018-12-10 16:27:41 (50.9 MB/s) - ‘City_Employee_Payroll.csv’ saved [115662426]



With a dataset over 110M, in order to save time here, we only look at the first 1000 rows here about its statistical features.

In [2]:
!head -n 1000 City_Employee_Payroll.csv |csvstat

  1. "Row ID"

	Type of data:          Text
	Contains null values:  False
	Unique values:         999
	Longest value:         21 characters
	Most common values:    3-1000027830ctFu (1x)
	                       3-1000155488ctFu (1x)
	                       3-1000194958ctFu (1x)
	                       3-1000198958ctFu (1x)
	                       3-1000232317ctFu (1x)

  2. "Year"

	Type of data:          Number
	Contains null values:  False
	Unique values:         1
	Smallest value:        2,013
	Largest value:         2,013
	Sum:                   2,010,987
	Mean:                  2,013
	Median:                2,013
	StDev:                 0
	Most common values:    2,013 (999x)

  3. "Department Title"

	Type of data:          Text
	Contains null values:  False
	Unique values:         40
	Longest value:         52 characters
	Most common values:    Police (LAPD) (231x)
	                       Water And Power (DWP) (181x)
	                       Recreation And Parks (124x)
	           

We can see our original dataset contains a relatively huge volumn of data and various columns. Therefore, we can have a quick and brief look at the complete dataset with xsv

In [3]:
!xsv stats City_Employee_Payroll.csv| xsv table

field                         type     sum                 min                                      max                                           min_length  max_length  mean                   stddev
Row ID                        Unicode                      3-1000027830ctFu                         8-999983159dwp972Fu                           12          21                                 
Year                          Integer  745198891           2013                                     2018                                          4           4           2015.6199717617972     1.7227487441516143
Department Title              Unicode                      Aging                                    Zoo                                           3           60                                 
Payroll Department            Integer  1381356220          101                                      8702                                          0           4           4591.374098830309      2532.79

From the original dataset, we can see we have some numbers and text with different columns, and the last column also contains a website link which calls for tidying up in the following procedures.

It seems like that from the above xsv step, we can only have a simple glance at the dataset. Therefore, we desire to take a deeper understanding of the dataset with the help of Spark.

In [4]:
import findspark

In [5]:
findspark.init()

In [6]:
from pyspark import SparkContext

In [7]:
spark = SparkContext(appName='project')

In [8]:
spark

<SparkContext master=local[*] appName=project>

In [9]:
from pyspark import SQLContext

In [10]:
sqlc = SQLContext(spark)

In [11]:
City_Employee_Payroll = sqlc.read.csv("City_Employee_Payroll.csv", header=True, inferSchema=True)

In [12]:
City_Employee_Payroll.count()

369712

From the count result, we can see we have in total 369712 rows in our original dataset. With such volumn, we are able to conclude with some reasonable conclusions through our exploration with the dataset.

In [13]:
City_Employee_Payroll.columns

['Row ID',
 'Year',
 'Department Title',
 'Payroll Department',
 'Record Number',
 'Job Class Title',
 'Employment Type',
 'Hourly or Event Rate',
 'Projected Annual Salary',
 'Q1 Payments',
 'Q2 Payments',
 'Q3 Payments',
 'Q4 Payments',
 'Payments Over Base Pay',
 '% Over Base Pay',
 'Total Payments',
 'Base Pay',
 'Permanent Bonus Pay',
 'Longevity Bonus Pay',
 'Temporary Bonus Pay',
 'Lump Sum Pay',
 'Overtime Pay',
 'Other Pay & Adjustments',
 'Other Pay (Payroll Explorer)',
 'MOU',
 'MOU Title',
 'FMS Department',
 'Job Class',
 'Pay Grade',
 'Average Health Cost',
 'Average Dental Cost',
 'Average Basic Life',
 'Average Benefit Cost',
 'Benefits Plan',
 'Job Class Link']

In [14]:
City_Employee_Payroll.printSchema()

root
 |-- Row ID: string (nullable = true)
 |-- Year: integer (nullable = true)
 |-- Department Title: string (nullable = true)
 |-- Payroll Department: integer (nullable = true)
 |-- Record Number: string (nullable = true)
 |-- Job Class Title: string (nullable = true)
 |-- Employment Type: string (nullable = true)
 |-- Hourly or Event Rate: double (nullable = true)
 |-- Projected Annual Salary: double (nullable = true)
 |-- Q1 Payments: double (nullable = true)
 |-- Q2 Payments: double (nullable = true)
 |-- Q3 Payments: double (nullable = true)
 |-- Q4 Payments: double (nullable = true)
 |-- Payments Over Base Pay: double (nullable = true)
 |-- % Over Base Pay: double (nullable = true)
 |-- Total Payments: double (nullable = true)
 |-- Base Pay: double (nullable = true)
 |-- Permanent Bonus Pay: double (nullable = true)
 |-- Longevity Bonus Pay: double (nullable = true)
 |-- Temporary Bonus Pay: double (nullable = true)
 |-- Lump Sum Pay: double (nullable = true)
 |-- Overtime Pay: 

In [15]:
City_Employee_Payroll.select('Year').distinct().show()

+----+
|Year|
+----+
|2018|
|2015|
|2013|
|2014|
|2016|
|2017|
+----+



Our dataset covers five years from 2013 to 2018.Therefore, we can find out if there is a trend over time in our targeted columns and whether there appears to be a decline or increase over the years.

In [16]:
City_Employee_Payroll.select('Department Title').distinct().count()

93

In [17]:
City_Employee_Payroll.groupBy('Department Title').count().orderBy("count",ascending=False).show(10)

+--------------------+-----+
|    Department Title|count|
+--------------------+-----+
|Water And Power (...|68853|
|       Police (LAPD)|68339|
|Recreation And Parks|39845|
|     Airports (LAWA)|19490|
|Public Works - Sa...|19128|
|         Fire (LAFD)|18768|
|              Police|13796|
|Transportation (L...|10421|
|    General Services|10185|
|             Library| 9690|
+--------------------+-----+
only showing top 10 rows



In [18]:
City_Employee_Payroll.select('Employment Type').distinct().show()

+---------------+
|Employment Type|
+---------------+
|      Per Event|
|      Part Time|
|      Full Time|
+---------------+



In [19]:
City_Employee_Payroll.describe('Hourly or Event Rate').show()

+-------+--------------------+
|summary|Hourly or Event Rate|
+-------+--------------------+
|  count|              348691|
|   mean|   39.44351116029949|
| stddev|  32.592627261910756|
|    min|                 0.0|
|    max|               900.0|
+-------+--------------------+



In [20]:
City_Employee_Payroll.describe('Projected Annual Salary','Q1 Payments','Q2 Payments','Q3 Payments','Q4 Payments').show()

+-------+-----------------------+------------------+----------------+------------------+------------------+
|summary|Projected Annual Salary|       Q1 Payments|     Q2 Payments|       Q3 Payments|       Q4 Payments|
+-------+-----------------------+------------------+----------------+------------------+------------------+
|  count|                 369712|            369712|          369712|            369712|            369712|
|   mean|      80976.84876752207|20692.892993844063|19820.4063211901|21288.029570556988|16209.364176710586|
| stddev|      35192.90947551841|15310.921811132941|14192.5390098486|15310.946105485953|14768.417425990729|
|    min|                    0.0|         -18304.17|       -35949.38|         -34122.58|         -30728.27|
|    max|              374211.36|         262966.09|       217157.93|         291317.36|         500930.74|
+-------+-----------------------+------------------+----------------+------------------+------------------+



In [21]:
City_Employee_Payroll.describe('Base Pay','Total Payments','Payments Over Base Pay','% Over Base Pay').show()

+-------+------------------+-----------------+----------------------+--------------------+
|summary|          Base Pay|   Total Payments|Payments Over Base Pay|     % Over Base Pay|
+-------+------------------+-----------------+----------------------+--------------------+
|  count|            369712|           369712|                174094|              174094|
|   mean|63586.541619963886|  78010.693062307|    14386.282860236513|0.001192631566854...|
| stddev| 40808.39654069189|53568.85135993702|    22258.758627601743|0.027670668625413116|
|    min|         -12922.52|        -30728.27|             -30878.27|              -10.19|
|    max|          360942.4|        585164.96|             435621.77|                2.36|
+-------+------------------+-----------------+----------------------+--------------------+



In [22]:
City_Employee_Payroll.select('Benefits Plan').distinct().show()

+-------------+
|Benefits Plan|
+-------------+
|          DWP|
|         null|
|       Police|
|         Fire|
|         City|
+-------------+



In [23]:
City_Employee_Payroll.select('Benefits Plan','Average Benefit Cost','Average Basic Life','Average Dental Cost','Average Health Cost').distinct().show()

+-------------+--------------------+------------------+-------------------+-------------------+
|Benefits Plan|Average Benefit Cost|Average Basic Life|Average Dental Cost|Average Health Cost|
+-------------+--------------------+------------------+-------------------+-------------------+
|         City|             12704.5|             10.59|             419.26|           12274.65|
|          DWP|            18755.63|               0.0|            1268.69|           17486.94|
|       Police|             14665.1|            193.96|             959.69|           13511.45|
|         Fire|            14073.96|             159.6|             889.32|           13025.04|
|       Police|            13042.56|            193.32|             873.24|            11976.0|
|          DWP|            17111.02|               0.0|            1250.49|           15860.53|
|          DWP|            19392.62|               0.0|            1248.19|           18144.43|
|          DWP|            18124.69|    

The records of "base pay" and "total payments" is 369712, but the records of "payments over base pay" and " % over base pay" is 174094. We can draw an conclusion that the last two columns seems to miss some data, so we have to deal with this problem if we want to conduct further analysis on these two variables.

---

Finally,from all the above result, we can see that for our analysis here, our original dataset contains some unnecessary columns and meanwhile, have null values. Therefore, in order to complete the following detailed analysis, we will tidy up our original dataset in the next section.

# Part 2 - Wrangling

Based on what you found above, wrangle the data into a format suitable for analysis. This may involve cleaning, filtering, merging, and modeling steps, any and all of which are valid for this project. Describe your process as you proceed, and document any scripts, databases, or other models you develop. Be specific about any key decisions to modify or remove data, how you overcame any challenges, and all assumptions you make about the meaning of variables and their values.
Verify that your wrangling steps have succeeded (for example, if you loaded the data into a dimensional model, ensure that the fact table contains the right number of records).

----

## 0 - Data preprocessing

On our first attempt, we try to wrangle with Trifacta directly. However, considering the volumn of our dataset, it is hard to fit it into it.Therefore, we conducted some simple data preprocessing steps with pandas. The detailed python code shows as below.

---

In the previous part, we delete some columns that are not related to our analysis here and unnecessary for our current exploration. Besides, after the python process, we will have a file with proper size and are able to fit the dataset into Trifacta for further wrangling steps now.

The detailed process receipt looks as below.

* replacemissing col: {Benefits Plan} with: custom text: 'Unknown'
* rename type: manual mapping: [column2,'ID']
* drop col: {Pay Grade} action: Drop
* replacemissing col: {Hourly or Event Rate} with: custom text: '0'
* replacepatterns col: ID, Year, {Department Title}, {Job Class Title}, {Employment Type}, {Hourly or Event Rate}, {Projected Annual Salary}, {Q1 Payments}, {Q2 Payments}, {Q3 Payments}, {Q4 Payments}, {Payments Over Base Pay}, {% Over Base Pay}, {Total Payments}, {Base Pay}, {Permanent Bonus Pay}, {Longevity Bonus Pay}, {Temporary Bonus Pay}, {Lump Sum Pay}, {Overtime Pay}, {Other Pay & Adjustments}, {Other Pay (Payroll Explorer)}, {Job Class}, {Average Health Cost}, {Average Dental Cost}, {Average Basic Life}, {Average Benefit Cost}, {Benefits Plan} with: ' ' on: '\"'
* replacemissing col: {Lump Sum Pay},{Overtime Pay},{Payments Over Base Pay},{% Over Base Pay} with: custom text: '0'

For our wrangling with Trifacta, we first replace some missing values in benefits plan with unknown. From the first section, we have already known that there are some categories within this column such as city and police. However, it will be hard to distinguish employee's benefit plan from their department or job titles. Therefore, it is reasonable for us to replace the missing values with unknown which also clearly illustrates the situation inside the dataset.

Then, as for the column2, it is a column generated from the python dataframe which is an unique identifier of each row. Therefore, we transform its name into ID to be a marker for every employee.After that we further drop some unnecessary columns that are irrelated to our analysis.

Last but not least, we replace the missing values in other five columns. There are some missing values with the hourly or event rate. We assume that the null value comes from statistical reasons. One possible explanation is those people simply get paid by year instead of by hour or event. Therefore, it is reasonable fore them to have blank rate here in this column. As for the other four columns, there may exist no over base or over time pay here within the dataset and leads to the nulls. Initially, we prefer to replace with some negative values in order to distinguish those missing values with other normal values we have. When we take a deeper look into our dataset, however, we can see there appears to have some negative values in some of the columns. The insertion of negative values will disturb the current distribution and make the picture no longer a reflection of the real situation. Therefore, we decide to simply replace those values with 0 in order to retain the balance with the picture.

## 1 - Examine the Wrangled Dataset

The file below is the wrangled dataset. People who read about this notebook will definitely get identical data file with our orignal dataset after the entire wrangling process documented in the last section.

In [24]:
!wget -O employee_2013_to_2018.csv https://s3.amazonaws.com/jiaruxu/employee_2013_to_2018.csv

--2018-12-10 16:28:43--  https://s3.amazonaws.com/jiaruxu/employee_2013_to_2018.csv
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.128.197
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.128.197|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 98855757 (94M) [text/csv]
Saving to: ‘employee_2013_to_2018.csv’

employee_2013_to_20 100%[===================>]  94.28M  38.3MB/s    in 2.5s    

2018-12-10 16:28:45 (38.3 MB/s) - ‘employee_2013_to_2018.csv’ saved [98855757/98855757]



Similar to the previous study, we decide to use spark to explore our wrangled data.

In [25]:
employee_2013_to_2018 = sqlc.read.csv("employee_2013_to_2018.csv", header=True, inferSchema=True)

In [26]:
employee_2013_to_2018.count()

369712

Luckily we have the exact same rows compared with our original dataset and the column also looks still in sequence.

In [27]:
employee_2013_to_2018.groupBy('Employment Type').count().orderBy("count",ascending=False).show()

+---------------+------+
|Employment Type| count|
+---------------+------+
|      Full Time|303539|
|      Part Time| 65142|
|      Per Event|  1031|
+---------------+------+



In [28]:
employee_2013_to_2018.select('Benefits Plan').distinct().show()

+-------------+
|Benefits Plan|
+-------------+
|          DWP|
|      Unknown|
|       Police|
|         Fire|
|         City|
+-------------+



Looks like we have successfully completed the replacement here and substitute the null value with our custom value here.

## 2 - Create Database and Load Data

Basing on our goals, we create a database here with star schema. We set the whole schema with three dimension tables and one fact table. Respectively, the three dimension tables measure about the year,department and employment situations. In our fact, we would like to target on various forms of payment,namely the total payment, base payment,overtime payment,permanent bonus payment and quarter payment.

In [29]:
from IPython.display import Image
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/final_project_schema.png?raw=true")  

In [30]:
%load_ext sql

In [31]:
!dropdb --if-exists -U student project

In [32]:
!createdb -U student project

In [33]:
%sql postgresql://student@/project

'Connected: student@project'

In [34]:
# load our original data into a table for further processing

In [35]:
%%sql
DROP TABLE IF EXISTS employee_info;

CREATE TABLE employee_info (
    ID INTEGER,
    year INTEGER,
    department_title VARCHAR(100),
    job_class_title VARCHAR(500),
    employment_type VARCHAR(500),
    hourly_or_event_rate NUMERIC(20,2),
    projected_annual_salary NUMERIC(20,2),
    Q1_payments NUMERIC(20,2),
    Q2_payments NUMERIC(20,2),
    Q3_payments NUMERIC(20,2),
    Q4_payments NUMERIC(20,2),
    payment_over_base_pay NUMERIC(20,2),
    percent_over_base_pay NUMERIC(20,2),
    total_payments NUMERIC(20,2),
    base_pay NUMERIC(20,2),
    permanent_bonus_pay NUMERIC(20,2),
    longevity_bonus_pay NUMERIC(20,2),
    temporary_bonus_pay NUMERIC(20,2),
    lump_sum_pay NUMERIC(20,2),
    overtime_pay NUMERIC(20,2),
    other_pay_adj NUMERIC(20,2),
    other_pay NUMERIC(20,2),
    job_class INTEGER,
    average_health_cost NUMERIC(20,2),
    average_dental_cost NUMERIC(20,2),
    average_basic_life NUMERIC(20,2),
    average_benefit_cost NUMERIC(20,2),
    benefits_plan VARCHAR(50)
)

 * postgresql://student@/project
Done.
Done.


[]

In [36]:
!pwd

/home/ubuntu


In [37]:
# please remember to modify the following path according to the previous result

In [38]:
%%sql
COPY employee_info FROM '/home/ubuntu/employee_2013_to_2018.csv'
CSV
HEADER;

 * postgresql://student@/project
369712 rows affected.


[]

In the following code, we make a few modification on the department title.

At first, we did not figure out whether there is any relationship between those similar department titles. We searched it online and did not have any valuable conclusions. Then, during one visualization of our data, we find out the similar forms in the following code denotes exactly the same thing. They are simply different names within the same department and in some year the statistical department decided to take a different name and causes our previous confusion. Therefore, we uniform their formats in order to help our following analysis.

In [39]:
%%sql
update employee_info
set department_title='Police'
where department_title='Police (LAPD)';

update employee_info
set department_title='Fire'
where department_title ='Fire (LAFD)';

update employee_info
set department_title='Harbor'
where department_title='Harbor (Port of LA)';

update employee_info
set department_title='Airports'
where department_title='Airports (LAWA)';

 * postgresql://student@/project
68339 rows affected.
18768 rows affected.
5988 rows affected.
19490 rows affected.


[]

In [40]:
%%sql
SELECT * FROM employee_info
limit 3;

 * postgresql://student@/project
3 rows affected.


[(3, 2013, 'Zoo', 'Light Equipment Operator', 'Full Time', Decimal('30.23'), Decimal('63126.77'), Decimal('15319.59'), Decimal('17762.14'), Decimal('15601.85'), Decimal('18233.32'), Decimal('5924.08'), Decimal('0.00'), Decimal('66916.90'), Decimal('60992.82'), Decimal('3446.79'), Decimal('1677.29'), Decimal('800.00'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('5924.08'), 3523, Decimal('10710.24'), Decimal('405.24'), Decimal('11.40'), Decimal('11126.88'), 'City'),
 (6, 2013, 'Public Works - Sanitation', 'Civil Engineering Associate II', 'Full Time', Decimal('45.60'), Decimal('95212.80'), Decimal('20937.60'), Decimal('24427.20'), Decimal('21729.60'), Decimal('25536.00'), Decimal('0.00'), Decimal('0.00'), Decimal('92630.40'), Decimal('92630.40'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), 7246, Decimal('10710.24'), Decimal('405.24'), Decimal('11.40'), Decimal('11126.88'), 'City'),
 (7, 2013, 'Water And Power (DWP)', 'Construction Helper', 'Full Time', Decimal('0.00'), Decimal('59278.32'), Decimal('19260.07'), Decimal('15752.00'), Decimal('0.00'), Decimal('0.00'), Decimal('13435.67'), Decimal('0.01'), Decimal('35012.07'), Decimal('21576.40'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('0.00'), Decimal('723.95'), Decimal('12711.72'), Decimal('12711.72'), 941, Decimal('15860.53'), Decimal('1250.49'), Decimal('0.00'), Decimal('17111.02'), 'DWP')]

Create database tables  based on the following schema, and load the dataset into it. 

* `years` `(year_key,year)`
* `department` `(department_key,department_title)`
* `employment` `(employment_key,job_class_title,employment_type)`
* `payment_fact` `(year_key,department_key,employment_key,total_payments,base_pay,q1_payments,q2_payments,`
 `q3_payments,q4_payments,permanent_bonus_pay,overtime_pay)`
                

### 2-1  Create Fact Table -- payment_fact

In [41]:
%%sql
DROP TABLE IF EXISTS payment_fact;

CREATE TABLE payment_fact (
    ID INTEGER,
    year INTEGER,
    department_title VARCHAR(100),
    job_class_title VARCHAR(500),
    employment_type VARCHAR(500),
    hourly_or_event_rate NUMERIC(20,2),
    projected_annual_salary NUMERIC(20,2),
    Q1_payments NUMERIC(20,2),
    Q2_payments NUMERIC(20,2),
    Q3_payments NUMERIC(20,2),
    Q4_payments NUMERIC(20,2),
    payment_over_base_pay NUMERIC(20,2),
    percent_over_base_pay NUMERIC(20,2),
    total_payments NUMERIC(20,2),
    base_pay NUMERIC(20,2),
    permanent_bonus_pay NUMERIC(20,2),
    longevity_bonus_pay NUMERIC(20,2),
    temporary_bonus_pay NUMERIC(20,2),
    lump_sum_pay NUMERIC(20,2),
    overtime_pay NUMERIC(20,2),
    other_pay_adj NUMERIC(20,2),
    other_pay NUMERIC(20,2),
    job_class INTEGER,
    average_health_cost NUMERIC(20,2),
    average_dental_cost NUMERIC(20,2),
    average_basic_life NUMERIC(20,2),
    average_benefit_cost NUMERIC(20,2),
    benefits_plan VARCHAR(50)
)

 * postgresql://student@/project
Done.
Done.


[]

In [42]:
%%sql
INSERT INTO payment_fact
SELECT * FROM employee_info

 * postgresql://student@/project
369712 rows affected.


[]

### 2-2  Create Dimensions Tables  -- years,department,employment,benefit_plan

#### 2-2-1  Create department table  -- department (department_key,department_title)

In [43]:
%%sql
DROP TABLE IF EXISTS department;

CREATE TABLE department (
    department_key SERIAL PRIMARY KEY,
    department_title VARCHAR(500));

 * postgresql://student@/project
Done.
Done.


[]

In [44]:
%%sql
INSERT INTO department (department_title)
SELECT DISTINCT department_title
FROM payment_fact;

 * postgresql://student@/project
89 rows affected.


[]

#### 2-2-2  Create employment table  -- employment (employment_key,job_class_title,employment_type)

In [45]:
%%sql
DROP TABLE IF EXISTS employment;

CREATE TABLE employment (
    employment_key SERIAL PRIMARY KEY,
    job_class_title VARCHAR(500),
    employment_type VARCHAR(100)
);

 * postgresql://student@/project
Done.
Done.


[]

In [46]:
%%sql
INSERT INTO employment (job_class_title, employment_type)
SELECT DISTINCT job_class_title, employment_type
FROM payment_fact;

 * postgresql://student@/project
1878 rows affected.


[]

#### 2-2-3  Create years table  -- years (year_key,year)

In [47]:
%%sql
DROP TABLE IF EXISTS years;

CREATE TABLE years (
    year_key SERIAL PRIMARY KEY,
    year INTEGER
);

 * postgresql://student@/project
Done.
Done.


[]

In [48]:
%%sql
INSERT INTO years (year)
SELECT DISTINCT year
FROM employee_info
ORDER BY year;

 * postgresql://student@/project
6 rows affected.


[]

#### 2-2-4  Create index to speed up

In [49]:
%%sql
DROP INDEX IF EXISTS idx_department;

CREATE INDEX idx_department ON department (department_title);

 * postgresql://student@/project
Done.
Done.


[]

In [50]:
%%sql
DROP INDEX IF EXISTS idx_employment;

CREATE INDEX idx_employment ON employment (job_class_title,employment_type);

 * postgresql://student@/project
Done.
Done.


[]

#### 2-2-5  Link dimensions to fact with Surrogate Keys

In [51]:
%%sql
ALTER TABLE payment_fact
ADD COLUMN year_key INTEGER,
ADD CONSTRAINT fk_year_key
    FOREIGN KEY (year_key)
    REFERENCES years (year_key);
    
ALTER TABLE payment_fact 
ADD COLUMN department_key INTEGER,
ADD CONSTRAINT fk_department_key
    FOREIGN KEY (department_key)
    REFERENCES department (department_key);
    
ALTER TABLE payment_fact
ADD COLUMN employment_key INTEGER,
ADD CONSTRAINT fk_employment_key
    FOREIGN KEY (employment_key)
    REFERENCES employment (employment_key);

 * postgresql://student@/project
Done.
Done.
Done.


[]

In [ ]:
%%sql
UPDATE payment_fact
SET year_key = years.year_key
FROM years
WHERE payment_fact.year = years.year;

UPDATE payment_fact
SET department_key = department.department_key
FROM department
WHERE payment_fact.department_title = department.department_title;

UPDATE payment_fact
SET employment_key = employment.employment_key
FROM employment
WHERE payment_fact.job_class_title = employment.job_class_title
AND payment_fact.employment_type = employment.employment_type;

 * postgresql://student@/project
369712 rows affected.
369712 rows affected.


In [ ]:
# drop unnecessary columns

In [ ]:
%%sql
ALTER TABLE payment_fact
DROP COLUMN id,
DROP COLUMN year,
DROP COLUMN department_title,
DROP COLUMN job_class_title,
DROP COLUMN employment_type,
DROP COLUMN hourly_or_event_rate,
DROP COLUMN projected_annual_salary,
DROP COLUMN payment_over_base_pay,
DROP COLUMN percent_over_base_pay,
DROP COLUMN temporary_bonus_pay,
DROP COLUMN lump_sum_pay,
DROP COLUMN other_pay_adj,
DROP COLUMN other_pay,
DROP COLUMN job_class,
DROP COLUMN average_health_cost,
DROP COLUMN average_dental_cost,
DROP COLUMN average_basic_life,
DROP COLUMN longevity_bonus_pay,
DROP COLUMN average_benefit_cost,
DROP COLUMN benefits_plan;

In [ ]:
# take a look at the fact table

In [ ]:
%%sql
select * from payment_fact
limit 5

# Part 3 - Analysis 

Explore and analyze your data in its wrangled form. Follow through on the themes you identified in Part 1 with queries or scripts that answer the questions you had in mind. Be clear about the answers you discover, discussing them and whether the results match your expectations. Include charts or other visuals that support your analysis. You may use Tableau, matplotlib, ggplot, or other tools we have not covered in class for visualization (and only for visualization), but be sure to export images from those tools and to include any images properly in your notebook writeup and slides.

In order to be more clear about our following analysis process, we here depict our analysis structure with the following picture.

In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/structure.jpg?raw=true")  

In [ ]:
%matplotlib inline

## a. by year

In [ ]:
%%sql
select years.year,round(avg(total_payments),2) as average_payment
from payment_fact
join years on years.year_key=payment_fact.year_key
group by years.year_key
order by years.year asc;

In [ ]:
_.bar()

Total payments in different year have not different changes,therefore year is the unnecessary variable to us to analyze the overall picture fluctuate through years.There is not a distincitve pattern over how the average payment changes, but in recent year, there is a clear drop in the average payment。

In [ ]:
%%sql
select  years.year,round(avg(base_pay),2) as average_base_pay
from payment_fact
join years on years.year_key=payment_fact.year_key
group by years.year_key
order by years.year asc
limit 10;

In [ ]:
_.bar()

In [ ]:
%%sql
select  years.year,round(avg(overtime_pay),2) as average_overtime_pay
from payment_fact
join years on years.year_key=payment_fact.year_key
group by years.year_key
order by years.year asc
limit 10;

In [ ]:
_.bar()

As for the average time pay, it increased from year 2013 to year 2017. Considering that currently we have no access to the data of the fourth quarter in 2018, we temporarily ignore the last column.

The overall changes seem to give little information, so we have to dig deeper to analyze the patterns between data.

## b.by department

After analyzing the overall fluctuations of average base pay, overtime payment and total payment from year 2013 to year 2018, we discovered that these differences between years are relatively small. Then we moved on to explore the data into details. 

This part is combined by two elements, which are the departments with highest average different types of payments and influences of different employment types on salaries in different departments.


### b-1 payment type

First of all, we tried to find out departments with highest average total payment, base payment and overtime payments. We also combined the influence of years to confirm the analysis in the overall year fluctuations and to eliminate the possibilities of results caused by outliers.

#### b-1-1 total pay

Firstly, we were curious about the departments with top 10 high payments.

In [ ]:
%%sql
select years.year,department.department_title,round(avg(total_payments),2) as average_total_payments
from payment_fact
join years on payment_fact.year_key=years.year_key
join department on department.department_key=payment_fact.department_key
group by department.department_title,years.year
order by avg(total_payments) desc
limit 10;

In [ ]:
_.bar()

In this table, we take the influence of years into consideration. We can see from this table, Public Accountability, Fire, City Attorney ranked top 3. Due to lack of detailed information, we decided to conduct further analysis on base payment and overtime payments for these departments later.

We observe that the Los Angeles Department of Convention and Tourism Development shows up in this chart. However, this department did not appear in our former analysis when we take the average of annual payment for each department, so we think a possible reason for this pattern could be tourism in LA reachs a peak in these two years. 

In [ ]:
import matplotlib.pyplot as plt

# Data to plot
labels = 'Fire', 'Public Accountability', 'Los Angeles Department of Convention and Tourism Development', 'City Attorney'
sizes = [6, 2, 1, 1]
colors = ['gold', 'yellowgreen', 'lightcoral', 'lightskyblue']
explode = (0.1, 0, 0, 0) 
plt.pie(sizes, explode=explode, labels=labels, colors=colors,
        autopct='%1.1f%%', shadow=True, startangle=140)
 
plt.axis('equal')
plt.show()


When we put the influence of years into consideration, the fire department occupied the largest proportion in the top 10 departments. The public accountability's average total payments in year 2017 and year 2018 are still ranked in top10. We were interested in this two departments’ outstanding performances. Next part, we conducted analysis on base pay.

#### b-1-2 base pay 

In [ ]:
%%sql
select department.department_title,round(avg(base_pay),2) as average_base_payment
from payment_fact
join department on department.department_key=payment_fact.department_key
group by department.department_title
order by avg(base_pay) desc
limit 10;

In [ ]:
_.bar()

From the above table, we find that Public Accountability's base pay is significantly higher than other departments, which agreed with our previous finding that this department has the highest average total payment. There are very few people who worked for this department, and we think it's because their job characteristics. 
On the other hand, fire department's average base pay is lower than city attorney and city administrative officer, while in previous analysis, fire's total payment is ranked at top 2. This draw us into thinking that the bonus (pay above base pay) plays a huge role in salary for Fire, similar to Police and Fire(LAFD), and there is a big difference between Fire and Police when we sort total payment. 


In [ ]:
%%sql
select years.year,department.department_title,round(avg(base_pay),2) as average_payment
from payment_fact
join years on payment_fact.year_key=years.year_key
join department on department.department_key=payment_fact.department_key
group by department.department_title,years.year
order by avg(base_pay) desc
limit 10;


We observe that the Los Angeles Department of Convention and Tourism Development shows up in this chart again in year 2016 and 2017, so we did a little background research that visitors in 2017 reached 48.3 millions (https://www.latimes.com/business/la-fi-la-tourism-20180110-story.html), a huge increase from previous years, which explains why this specific department has a particular high payment in this calendar year.




In [ ]:
# Data to plot
import matplotlib.pyplot as plt
labels = 'City Attorney', 'Public Accountability', 'Los Angeles Department of Convention and Tourism Development', 'City Administrative Officer'
sizes = [5, 2, 2, 1]
colors = ['gold', 'yellowgreen', 'lightcoral', 'lightskyblue']
explode = (0.1, 0, 0, 0) 
plt.pie(sizes, explode=explode, labels=labels, colors=colors,
        autopct='%1.1f%%', shadow=True, startangle=140)
 
plt.axis('equal')
plt.show()


We also took the influence of years into consideration. The results of this table are slightly different from the former table. Actually, city attorney ranked in top10 in 2013,2014,2015,2016 and occupied the largest proportion. Public accountability occupied 1st and 2nd place in this table and this phenomenon is consistent with the former result.

After analyzing the base payment, we explored the fluctuations of overtime payments.


#### b-1-3 overtime pay

Have already explored the average base pay and total payments, we moved on to the overtime payments.

In [ ]:
%%sql
select department.department_title,round(avg(overtime_pay),2) as average_overtime_payment
from payment_fact
join department on department.department_key=payment_fact.department_key
group by department.department_title
order by avg(overtime_pay) desc
limit 10;

In [ ]:
_.bar()

Surprisingly, Fire has a significantly higher overtime pay than the other department, almost four times as the water and power. While police does not have a very high over time pay, which explains why in total payments, it is lower than Fire department. Another interesting finding is the top 10 department are all related with service job.

In [ ]:
%%sql
select years.year,department.department_title,round(avg(overtime_pay),2) as average_overtime_payment
from payment_fact
join years on payment_fact.year_key=years.year_key
join department on department.department_key=payment_fact.department_key
group by department.department_title,years.year
order by avg(overtime_pay) desc
limit 10;


In [ ]:
year1 = [2013, 2014, 2015, 2016, 2017, 2018]
data1 = [39534.75, 45772.91,50618.16 , 47914.62, 49943.76, 37388.08]
year2 =[ 2015, 2016,2017, 2018] 
data2=[ 14906.04, 14796.30,14996.30, 13914.53]
plt.plot(year1, data1, color='g',label="Fire")
plt.plot(year2, data2, color='orange',label="Water And Power (DWP)")
plt.legend()
plt.xlabel('year')
plt.ylabel('average_overtime_payment')

Next, we explored this problem into details based on different years. The results are completely the same to the former table. After some survey, we found out that the high overtime payment of these two department are determined by an important factor 24/7 operation hours. We also associated this phenomenon with the California's weather and frequent mountain fires, which requires fire department to make contributions. Dry and hot weather also need the efforts of the department of water and power.

Also, now we can answer the question why fire department with relatively low base payment can have high total payments. The overtime payments make great contributions on the salary.


### b-2 employment type

Secondly, we conducted analysis about average base pay, overtime payment and total payment of different departments based on different employment types.

In [ ]:
%%sql
select employment.employment_type,round(avg(total_payments),2) as average_payment,round(avg(base_pay),2) as average_base_pay,round(avg(overtime_pay),2) as average_overtime_pay
from payment_fact
join employment on employment.employment_key=payment_fact.employment_key
group by employment.employment_type
order by avg(total_payments) desc;

In [ ]:
type1=['full time','part time','per event']
data1=[92844.28,75388.67,10325.24]
data2=[10048.41,9521.76,37.08]
data3=[4904.18,4887.43,0.00]
plt.plot(type1, data1, color='g',label="average_payment")
plt.plot(type1, data2, color='orange',label="average_base_pay")
plt.plot(type1, data3, color='red',label="average_overtime_pay")
plt.legend()
plt.xlabel('type')
plt.ylabel('payment')

At first, let us take a first glance of the overall distribution of these payments. There is no doubt that full time employees have the highest payments in all categories, and part-time employees ranked 2nd.


Then we explored the data into details based on different department.

#### b-2-1 full-time

In [ ]:
%%sql
select department.department_title,round(avg(total_payments),2) as average_payment
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where employment.employment_type = 'Full Time'
group by department.department_title
order by avg(total_payments) desc
limit 10;

In [ ]:
_.bar()

Similar to former results, public accountability and fire department have the largest average total payments. Departments associated with social service offering high payments to their employees.

#### b-2-2 part-time

Then we analyzed the role part time employees played in different departments.

In [ ]:
%%sql
select department.department_title,round(avg(total_payments),2) as total_payment
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where employment.employment_type = 'Part Time'
group by department.department_title
order by avg(total_payments) desc
limit 10;

In [ ]:
_.bar()

We noticed that City Attorney, Office of the Chief Legislative Analyst , different council offered high average payments to their part-time employees. We conducted some research on these departments and found out they all related to laws and their have some interrelation between these departments. For example, City attorneys must hold a doctoral degree in law. They must also be licensed to practice law in the state where the city is located. And the City attorneys should give advice to council members. We assumed that some city attorney should be someone out of the government system in order to give wiser advice, so they hire part-time employees.

#### b-2-3 per event

In [ ]:
%%sql
select department.department_title,round(avg(total_payments),2) as avg_payment
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where employment.employment_type = 'Per Event'
group by department.department_title
order by avg(total_payments) desc
limit 10;

Finally, we found out that cultural affairs department tends to pay highest average payment to its per event employees. Employment relations board ranked 2nd. Also Recreation And Parks department ranked in top 10. We guessed that the main reason behind this phenomenon is these department do not give service to public routinely and only held big events.

## c - a deeper analysis over overtime payment across four departments

### c-1 Fire

First, we want to see the effect of employment type on salary performance in fire department.

In [ ]:
%%sql
select employment.job_class_title,round(avg(overtime_pay),2) as overtime, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Part Time' and department.department_title like '%Fire%')
group by employment.job_class_title
order by avg(overtime_pay) desc
limit 10;


In [ ]:
%%sql
select employment.job_class_title,round(avg(overtime_pay),2) as overtime, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Per Event' and department.department_title like '%Fire%')
group by employment.job_class_title
order by avg(overtime_pay) desc
limit 10;


From the result above, we can see that the overtime payment does not apply to these two employment types. So we decide to focus on full-time employment in fire department.

In [ ]:
%%sql
select employment.job_class_title,round(avg(overtime_pay),2) as overtime, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Full Time' and department.department_title like '%Fire%')
group by employment.job_class_title
order by avg(overtime_pay) desc
limit 10;


In Fire department, Fire Helicopter Pilot has the highest overtime pay, and the fireboat pilot has second highest overtime pay, it is reasonable since LA has a lot wide forest fire through out the year, helicopter is a major way to use in these cases. What’s more interesting is the overtime pay for fire helicopter pilot V and fireboat pilot  is even higher than their base payment. 

### c-2 Police

In [ ]:
%%sql
select employment.job_class_title,round(avg(overtime_pay),2) as overtime, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Full Time' and department.department_title like '%Police%')
group by employment.job_class_title
order by avg(overtime_pay) desc
limit 10;


On the other hand, police has a general a lot lower overtime pay than Fire, the highest job title is police captain, followed by police surveillance specialist II, and general automotive supervisor. It is a little surprising to see principal fingerprint identification expert I ranked at NO.4 as it is an assistance position.

### c-3 Water and Power

In [ ]:
%%sql
select employment.job_class_title,round(avg(overtime_pay),2) as overtime, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Full Time' and department.department_title like '%Water%')
group by employment.job_class_title
order by avg(overtime_pay) desc
limit 10;

Although Water and Power department ranked at NO.3 for overtime pay, its base pay is very low compare to fire and police, so it's total pay is lower than the other two groups, thus did not show in the rank for average total payment.

### c-4 Public Accountability

In [ ]:
%%sql
select employment.job_class_title,round(avg(overtime_pay),2) as overtime, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Full Time' and department.department_title like '%Public Accountability%')
group by employment.job_class_title
order by avg(overtime_pay) desc
limit 10;


Public Accountability gets to rank at top 1 really because it has a high base pay.

### Permanent bonus 

In [ ]:
%%sql
select employment.job_class_title,round(avg(permanent_bonus_pay),2) as permanent, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Full Time' and department.department_title like '%Fire%')
group by employment.job_class_title
order by avg(permanent_bonus_pay) desc
limit 10;

In [ ]:
%%sql
select employment.job_class_title,round(avg(permanent_bonus_pay),2) as permanent, 
round(avg(base_pay),2) as base_pay,round(avg(total_payments),2) as total
from payment_fact
join department on department.department_key=payment_fact.department_key
join employment on employment.employment_key=payment_fact.employment_key
where (employment.employment_type = 'Full Time' and department.department_title like '%Police%')
group by employment.job_class_title
order by avg(permanent_bonus_pay) desc
limit 10;

Both Fire and Police departments shows more advanced job title tend to have a higher permanent bonus pay. So there isn't many we can analyze through this lens. And the amount is generally slightly lower for Police than Fire department, which again shows the total payment for Fire would be higher than Police.

## d. Seasonality & Position

After exploring the relationship between government payment with years,respective departments and job titles, we would prefer to probe deeper into the seasonality pattern and hierarchy relationships.With the former analysis, we still decide to focus on the following four departments:'Police','Fire','Public Accountability' and 'Water and Power' for this part.

In order to fulfill our goal, in the following steps, we will first group by department and then focus on our distinctive targets.

### d-1   discover the seasonality parttern across department

First, we would like to see if there is any pattern over seasonality over five years all through 2013 to 2018.

In [ ]:
%%sql
SELECT  department.department_title,
        round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join department on department.department_key=payment_fact.department_key
INNER JOIN
        (
            SELECT  'Police' col
            UNION SELECT  'Fire'
            UNION SELECT  'Public Accountability'
            UNION SELECT  '%DWP%'
            

        ) List ON department.department_title LIKE List.Col
    
group by department.department_title;

From the result above, we can see that in all out selected four departments, the payments in quarter 4 all drop dramatically. In order to figure out the reason behind this peculiar result, we tried to think of the possible reasons. However, hard to make a reasonable explanation. Therefore, we decide to take a further look into our original dataset and it turns out to be that there are many incomplete record for the fourth quarter in 2018 with 0 instead which drag the average result behind.

With this understanding, we decide to make a year filter in all following steps in order to avoid the influence of incomplete records.

In [ ]:
%%sql
SELECT  department.department_title,
        round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
INNER JOIN
        (
            SELECT  'Police' col
            UNION SELECT  'Fire'
            UNION SELECT  'Public Accountability'
            UNION SELECT  '%DWP%'
            

        ) List ON department.department_title LIKE List.Col
WHERE years.year in (2013,2014,2015,2016,2017)
    
group by department.department_title;

After adjusting the year, we find out in our targeted four departments, payments in fire, police and public accountability all peak at the third quarter and for water and power department, it reaches its highest in the first quarter.

First, for fire department, the average quarter payment rises steady from the first quarter to the third quarter and then experience a slightly drop in the last quarter. Combined with our former analysis that the component of payment in fire department is basically half the base-pay and half the overtime payment. We can reasonably claim that the rises partly comes from the more overtime payment in the fire department. In another words, it denotes that with the temperature goes higher, more fire-related incidents occur and correspondingly, more response from the fire department.

Then, for police department, its payment reaches the highest in the third quarter with the other three quarter approximately the same. For public accountability department, the payment attains its peak in the third quarter. 

What also appealing to us is the payment pattern in the DWP department. It has a different peak compared with the other three departments. Employees in Water and Power department have the highest average payment in the first quarter. It makes sense when we consider the local weather situation. Los Angeles has an extremely large volume of rain falls on January and February, which will inevitably lead to increase in water and power related incidents and thus employees get more subsidy. Besides, as we research into the source of fund into DWP, we find that the operations of this department are financed solely by the sale of water and electric services. Therefore, the more freezing the weather, the more usage on water and power resources by residents and business and more sales to the department. With a higher sale, the payment to employees would increase accordingly.

In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/seasonality_department.png?raw=true")  

### d-2 discover the seasonality parttern across positions

Consider the format of job titles in our dataset, it is complicated to extract the exact job positions. Therefore, we decide to discuss several most frequently appeared job positions here, namely Supervisor,Officer,Sergeant,Detective,Analyst,Typist,Lieutenant,Commander,Mechanic and Clerk.

In [ ]:
%%sql
SELECT  'Supervisor' as job_title,round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Supervisor%'
and  years.year in (2013,2014,2015,2016,2017)


union

SELECT  'Officer',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Officer%'
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Sergeant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Sergeant%'
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Detective',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Detective%'
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Analyst',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Analyst%'
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Typist',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Typist%'
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Lieutenant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Lieutenant%'
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Commander',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Commander%'
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Mechanic',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Mechanic%' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Clerk',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Clerk%' 
and  years.year in (2013,2014,2015,2016,2017)


In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/Seasonality_over_ten_job_positions.png?raw=true")  

From the above table and picture, we can see that across different departments, employees with different job titles display various payment patterns.

First, with average lower total payment, there is meanwhile also lower variation in the quarter payment. With a higher total payment, the variation at the same time lifts. For example, fore clerk and typist, the quarter payment remains approximately at the same level. On the opposite, for the sergeant and commander, there displays a quarterly fluctuation. Take the commander as an example, its quarter payment reaches its highest at the third quarter. 

When thinking of the reasons behind, we cannot deny that the base number matters. With a higher payment, the absolute variation is meanwhile larger. However, other than this, we can reasonably claim that it is the job feature that decides. Recalling our previous conclusion that almost every department has its payment peak at the third quarter, we can conclude with that the increase inthe quarter payment is from the lifted payment in higher job position.

Besides, mechanics and supervisors display a different feature than the others. However, we are curious about whether these job positions will display the same pattern in every department or they are strongly department-related. Therefore, in the following steps, we would like to probe deeper into the relationship between the job positions in three particular departments, that is the LAPD, LAFD and PWD.

### d-2-1 discover the seasonality parttern across positions in LAPD 

In [ ]:
%%sql
SELECT  'Supervisor' as police_title,round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Supervisor%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)

union

SELECT  'Officer',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Officer%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Sergeant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Sergeant%' and department.department_title = 'Police'  
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Detective',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Detective%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Analyst',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Analyst%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Typist',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Typist%' and department.department_title = 'Police'  
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Lieutenant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Lieutenant%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Commander',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Commander%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Mechanic',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Mechanic%' and department.department_title = 'Police'  
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Clerk',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Clerk%' and department.department_title = 'Police' 
and  years.year in (2013,2014,2015,2016,2017)


In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/LAPD.png?raw=true")  

From the figure above, we can see there is a distinctive and clear hierarchy of payment in the police department. In general, people with higher title, get a higher pay. For example, commanders are rewarded with the highest payment and the volume of payment follows a steady decline with Lieutenant, Sergeant and Supervisor in sequence. The payment within these job titles all reaches top in the third quarter.

With such an interesting finding, we try to propose the following possible reasons. First, the Police Department may receive more fund in the third quarter. With more fund, the department would be able to reward its employees with higher payment. Second, another possible reasons may lie in that the third quarter is a travel season. Therefore, a lot more people come to Los Angeles. With a denser population, the crime rate correspondingly rises. Therefore, having to deal with more cases, the police department have more payment as reward.

### d-2-2 discover the seasonality parttern across positions in Fire Department

In [ ]:
%%sql
SELECT  'Supervisor' as fire_title,round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Supervisor%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


union

SELECT  'Officer',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Officer%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Sergeant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Sergeant%' and department.department_title = 'Fire'
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Detective',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Detective%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Analyst',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Analyst%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Typist',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Typist%' and department.department_title = 'Fire'  
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Lieutenant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Lieutenant%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Commander',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Commander%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


union
SELECT  'Mechanic',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Mechanic%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Firefighter',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Firefighter%' and department.department_title = 'Fire' 
and  years.year in (2013,2014,2015,2016,2017)


In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/LAFD_.png?raw=true")  

From the above graph, we can see that obviously firefighters have the highest pay. In detail, the payment for firefighter rises with the local temperature. The higher the temperature, the higher the payment for firefighters. Recalling our previous result that the payment of fire department is in a relative large extent composed by over-time payment, therefore, we can reasonably conclude that it is the higher the temperature, the more the incidents and as a result, the more response from the fire department, which will finally lead to higher payment fore firefighters.

### d-2-3 discover the seasonality parttern across positions in Water and Power Department

In [ ]:
%%sql
SELECT  'Supervisor' as DWP_title,round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Supervisor%' and department.department_title = '%DWP%' 
and  years.year in (2013,2014,2015,2016,2017)

union

SELECT  'Officer',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Officer%' and department.department_title like '%DWP%' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Sergeant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Sergeant%' and department.department_title like '%DWP%'  
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Detective',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Detective%' and department.department_title like '%DWP%' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Analyst',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Analyst%' and department.department_title like '%DWP%' 
and  years.year in (2013,2014,2015,2016,2017)

union
SELECT  'Typist',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Typist%' and department.department_title like '%DWP%'  

union
SELECT  'Lieutenant',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Lieutenant%' and department.department_title like '%DWP%' 

union
SELECT  'Commander',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Commander%' and department.department_title like '%DWP%' 

union
SELECT  'Mechanic',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Mechanic%' and department.department_title like '%DWP%'  

union
SELECT  'Clerk',round(avg(q1_payments),2) as q1,round(avg(q2_payments),2) as q2,
        round(avg(q3_payments),2) as q3,round(avg(q4_payments),2) as q4,
        round(avg(total_payments),2) as total 
FROM    payment_fact 
join employment on employment.employment_key=payment_fact.employment_key
join department on department.department_key=payment_fact.department_key
join years on payment_fact.year_key=years.year_key
where employment.job_class_title like '%Clerk%' and department.department_title = '%DWP%' 
and  years.year in (2013,2014,2015,2016,2017)


In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/DWP.png?raw=true")  

With the above picture, we can clearly see there is an extremely high column with mechanic, which is totally different from the pattern revealed from the other two groups. Although this result seems surprising, we can easily make sense when considering its department features. In the department of water and power, there is a lot of work that is technical and a lot of experience is required for this particular position. Meanwhile, the working environment for this position is also not that comforting which also makes mechanics deserve higher payment.

As for the trend over mechanics' payment, we can see there is a clear decline over the first quarter to the last quarter. Basing on the picture here and the local climate situation, we infer that it is the weather in Los Angeles that leads to this drop. During the first quarter, there is a lot of rainfalls in LA, and floods are also more likely to happen, which calls for the need of mechanics from the water and power department to save the situation. This kind of emergency in return makes those mechanic welcomes a higher payment in the first quarter.

### Conclusion wrap up answers to questions in part 1

First focus answers:
1.    There should be a slowly increasing in salary over years due to inflation and living expense increase, but there might also not change a lot due to increasing demand. Does year really have an effect on salary?

         *A: Surprisingly, there is no big difference of salary over years, so year does not really have an effect on salary*
         
2.    What are the departments with the highest average total payment and if there is something in common through these departments? The most common city employees we know for daily are fireman and policeman, since they are service-based job, we think they probably have higher payments.

         *A: Public Accountability, Fire, City Attorney ranked top 3.  We find Fire department takes a huge portion of top 10 over 2013-2018, while Public Accountability only has data for 2016 and 2017, at this point, we cannot say if there is a common thing among these three departments. But our hypothesis that fireman has a higher salary, but police did not rank the top in our chart, so it looks like the police do not get paid very well.*

3.    We know that total payment is generated from base pay, overtime pay, permanent bonus pay, etc., we want to see if these bonus pays will make a difference in total pay?

         *A: we find that Public Accountability's base pay is significantly higher than other departments, which agreed with our previous observation that Public Accountability ranks top 1 for total payment. However, Fire department only ranked at 4th, so we figured there should be other reason that makes Fire department’s total payment very high. When we analyze the overtime pay, we surprisingly find that Fire has a significantly higher overtime pay than the other department, almost four times as the 2nd department the water and power. Later, we find that for Fire department, certain position has higher overtime pay than its base pay, and they generally have very high overtime pay than other departments, which make sense why even Fire has a lower base pay, it still has the highest total pay. For the permanent bonus, it has to do more with the job title, the more senior job like Fire Assistant Chief and Executive Director Police Commission has a lot of permanent bonus than entry-level positions. So digging into the constitute part of the total payment, the bonus does make a difference in total pay.*

4.    We predict that job title has an impact on salary, as people with higher positions will have a higher pay, but is there a big difference between different departments? We would want to make a comparison here.

         *A: We compared Fire, Police, Water and Power, and Public Accountability department. We find that overall, senior levels position like Fire Captain, Police Captain, and Senior Load Dispatcher do have a higher pay, however, there are exceptions. For the Fire department, Fire Helicopter Pilot and Fireboat Pilot have the highest salary, probably because of the techniques required for these types of jobs. Across departments, the Fire department has the highest average payment, while Police and Water and Power department are about the same. We also take a look at Public Accountability since it has the highest total pay, and find that they basically only have a base pay, and it is a lot higher than the others. We searched the duty of this department and found that they served as independent watchdog which means they monitor the performance of DWP department. So we conclude this department is a special case as the employees in this department need to really get paid well to do a good job on monitoring.*

5.    Our dataset also contains employment type, we should get a conclusion that full-time gets to pay most, but we probably would find some interesting thing if we categorize each employment type and the group of departments appears to have the highest pay might be different.

        *A: We indeed find full-time has the highest average payment. For Full-time group, we find half departments are service based such as Fire, Water and Power. The other half are advanced administration department, such as the City Administrative Officer (CAO), City Administrative Officer. For the Part-time group, a lot of these departments are for public facilities, such as Recreation and Parks, library, and transportation. For the Pre-event group, like it is categorized, most of the departments in the rank are seasonal jobs, like Cultural Affairs, pension (LAFPP) and Personnel. Overall, the only department that overlaps in all three employment types is Police, which is reasonable since Police would need more police force for certain period, but they do not always need a lot full-time police.*

For our second focus, we want to answer the following questions:
1.    We want to see if there is a difference in quarterly payment. We predict that for different department, there is a different pattern due to the department characteristics. 

        *A: we find out in our targeted four departments, payments in fire, police and public accountability all peak at the third quarter and for water and power department, it reaches its highest in the first quarter. Both Police and Fire have a steady increase for the first three quarters. While for the Water and Power department, it has the highest average payment in the first quarter. And we find a reasonable causation for this is the operations of this department are financed solely by the sale of water and electric services. So more usage of energy and water in cold season will reflect on a higher quarter salary.*

2.    Second, for different positions in some specific department, we wonder if there is also a pattern over how employees are rewarded through the years in different quarters. We predict that in different quarters of the year, there exists some variation in the pattern due to the seasonality and the feature of a specific department.

        *A: as we probe deeper into the fire, police and water and power department, we find out that in different departments, the payment pattern associated with quarter and position varies. In police department, there is a clear hierarchy over quarter payment. With higher title, the payment gets higher. In fire department, firefighters get the most payment instead of positions like supervisors. Besides, in Los Angeles Water and Power department(LAWPD),the mechanic ranks top when considering its quarter payment and followed by analysts. This pattern denotes that those two positions take up dominant positions in LAPWD. Possible explanation could be that the works in LAPWD is conducted mainly by those two types of workers. For mechanics, they are mainly responsible for conducting the detailed construction work and solving technical problems. For analyst, they will discover the systematical pattern and aim to raise the overall performance of the department. In another words, the mechanics emphasize front-end works and for analysts, they emphasizes on back-end works, and they altogether occupy a significant position inside the department.*

3.    Third, if we find out some seasonal difference after the previous step, we would like to explore further into the reason behind and find out how the variation differs among various positions in the same department and how large the variation is. 

        *A: Especially, in the fire department, the firefighters not only have the highest payment in total but also display a seasonal pattern across quarters. The quarter payment for firefighters rises with the temperature. The higher the temperature in that quarter, the more payment towards firefighters. Therefore, we propose a possible explanation for that with higher temperature in Los Angeles, more fire-related incidents may happen and correspondingly calls for more worktime from firefighters. Besides, because of the special location of Los Angeles, there may be more forest fires in summer which also add firefighters' workload. Then as for the Water and Power department, we can see that the payment for mechanics reach its peak in the first quarter which could be partially due to higher rainfalls in the first quarters. Heavy rainfalls would threat the water pipes and also the electrical facilities in the city and, as a result, will lead to more repairing work for the mechanics.*



# Bonus - Augment

Sometimes the most value can be gained from one dataset when it is studied alongside data drawn from other sources. Identify and describe at least one additional data source that can complement your analysis. Pull this additional data into your chosen environment and explore at least one more theme you are able to further analyze that depends upon a combination of data from both sources.

From the previous analysis section, we discover that there is a strong seasonality pattern over the payment in the fire department,especially in the payment of firefighters. Combined with our finding in part b that fire department employees' are paid with approximately half the base pay and half the overtime pay, we can easily infer that the increase in payment is due to the increase in incidents.

Therefore, in order to prove our hypothesis, we get the data from the Los Angeles Fire Department that describes about the responses in all fire stations grouped by month. With this dataset, we would like to test our previous conclusions and verify the hypothesis.

In [ ]:
!wget -O fire.csv https://s3.amazonaws.com/jiaruxu/All_Stations_Response_Metrics.csv

In [ ]:
!csvcut -n fire.csv

From the column,we can see the dataset has separate the incident into two categories. One concerns EMS situations and another talks about Non-EMS situations.For here, EMS means combined system, which uses the fire department for emergency response and a private or “third service” (police, fire, EMS) provider for transportation support.More than 85% of the LAFD's daily emergency responses are related to Emergency Medicine Services. 

To conduct our analysis, we create the following table.

In [ ]:
%%sql
DROP TABLE IF EXISTS fire_response;

CREATE TABLE fire_response (
    year VARCHAR(500),
    district VARCHAR(500),
    month VARCHAR(100),
    turnout VARCHAR(100),
    turnout_ems VARCHAR(100),
    travel VARCHAR(100),
    travel_ems VARCHAR(100),
    incident integer,
    incident_ems integer        
);

In [ ]:
!pwd

Also,please change the path according to the previous route.

In [ ]:
%%sql
COPY fire_response FROM '/home/ubuntu/fire.csv'
CSV
HEADER;

In [ ]:
%%sql
select year,count(*) from fire_response
group by year

From this table, we can see clearly, the data for 2017 is not complete. Therefore, we decide to only focus on the year from 2013 to 2016.

In [ ]:
%%sql
select year,sum(incident) as incident,sum(incident_ems) as ems_incident 
from fire_response
where year in ('2013','2014','2015','2016')
group by year
order by year;

From the table, we can see there is a clear lifting trend in each categories over the four years. There is a slight drop in 2016 for non-ems incident but in the whole picture, it is still an increasing trend.

In [ ]:
%%sql
select year,'q1' as season,sum(incident) as incident,sum(incident_ems) as ems_incident 
from fire_response
where month in ('Jan','Feb','Mar')
and year in ('2013','2014','2015','2016')
group by year
order by year;

In [ ]:
%%sql
select year,'q2' as season,sum(incident) as incident,sum(incident_ems) as ems_incident 
from fire_response
where month in ('Apr','May','Jun')
group by year
order by year;

In [ ]:
%%sql
select year,'q3' as season,sum(incident) as incident,sum(incident_ems) as ems_incident 
from fire_response
where month in ('Jul','Aug','Sep')
group by year
order by year;

In [ ]:
%%sql
select year,'q4' as season,sum(incident) as incident,sum(incident_ems) as ems_incident 
from fire_response
where month in ('Oct','Nov','Dec')
group by year
order by year;

In [ ]:
%%sql
select month,round(avg(incident),2) as incident,
            round(avg(incident_ems),2) as ems_incident,
            round(avg(incident),2)+round(avg(incident_ems),2) as total_incident
from fire_response
where month in ('Dec','Jan','Aug','Oct','Jul','Sep','May','Mar','Nov','Jun','Apr','Feb')
group by month
order by round(avg(incident),2)+round(avg(incident_ems),2) desc;

In [ ]:
Image(url="https://github.com/jiaruxu233/istm6212/blob/master/fire_response.png?raw=true")  

From the tables above, we can see there is not a clear overall seasonality pattern over fire incidents, opposite to what we originally expect. 

When thinking of the reasons behind, we consider that the way of how fire incidents are count may be taken into account. For example, both forest fire and fire alarm will be counted once into the incidence report. However, obviously, those two different type of fire incident have comparably distinctive impact. In order to put off a forest fire, more firefighters have to take part in and meanwhile a lot more resources are consumed.Inevitably, it will cause more over-time payment for the fire department than the incident caused by a false fire alarm.

Taking a deeper look into the situation, we can see two incident types display different patterns. First, for ems incident, there is not a clear seasonality over the years which denotes that the cooperation between fire department and other government department remains relatively constant over the months. Second, for non-ems situation, which reflects about real fire-related situations displays a month-based fluctuations. Non-EMS situations peak at December and January and remain roughly the average in all the other months. The rule revealed from this picture also goes against our original assumption.

In conclusion, no matter through the type of incidents or through the month over the years, we do not observe a clear trend as we expected. We raise one possible explanation here as the count of incidents can not be a true reflection about the workload of the fire department.For example, in some particular month, there may happens severe fire situations occasionally while in some other months, more incidents with less impact take place.

## Performance:

All memebers have contributed substantially and made 100% devotion to our project.

* Jiaru Xu : Write the draft version of our project, wrangle the dataset,conduct the analysis in part d and bonus part and make some modification about our whole notebook.

* Min Yang(Wendy) : Illustrate about the selection of dataset, the conclusion of our analysis,conduct the analysis in part b with Conying, provide some enlightening thoughts with our findings and examine about all the the code and the conclusions.

* Congying Ma : Conduct the analysis with Wendy in part b, draw the visualizations in part b and actively participate throughout our project. 

* Jiaying Liu : Draw our presentation slides and analyze about the year.